packages/dev/core/src/Materials/Textures/hdrCubeTexture.ts

```diff
@@ -278,7 +279,52 @@ export class HDRCubeTexture extends BaseTexture {
             const previousOnLoad = this._onLoad;
             const hdrFiltering = new HDRFiltering(engine);
             this._onLoad = () => {
-                hdrFiltering.prefilter(this, previousOnLoad);
+                let irradiancePromise: Promise<Nullable<BaseTexture>>;
+                let cdfGeneratedPromise: Promise<void>;
+                const cdfGenerator = this.getScene()?.iblCdfGenerator;
+                if (!this._generateHarmonics) {
+                    if (cdfGenerator) {
+                        // If we're using CDF maps, the importanceSamplingRenderer needs this texture to be
+                        // ready before it can generate the CDF maps and that won't happen until prefiltering
+                        // is done. So, lets make a new texture with the non-prefiltered data and set it as the
+                        // iblSource for the importanceSamplingRenderer. Then, we'll wait for that to be ready
+                        // before we continue with the prefiltering.
+                        cdfGenerator.iblSource = new BaseTexture(this._engine, this.getInternalTexture());
+                        cdfGeneratedPromise = new Promise((resolve) => {
+                            cdfGenerator.onGeneratedObservable.addOnce(() => {
+                                const oldTexture = cdfGenerator.iblSource;
+                                oldTexture?.dispose();
+                                cdfGenerator.iblSource = null;
+                                resolve();
+                            });
+                        });
+                    } else {
+                        cdfGeneratedPromise = Promise.resolve();
+                    }
+                    const hdrIrradianceFiltering = new HDRIrradianceFiltering(engine);
+                    irradiancePromise = cdfGeneratedPromise.then(() => hdrIrradianceFiltering.prefilter(this));
+                } else {
+                    irradiancePromise = Promise.resolve(null);
+                }
+                irradiancePromise.then((irradianceTexture) => {
+                    return hdrFiltering.prefilter(this).then(() => {
+                        if (!this._generateHarmonics && irradianceTexture) {
+                            this.irradianceTexture = irradianceTexture;
+                            const scene = this.getScene();
+                            if (scene) {
+                                scene.markAllMaterialsAsDirty(Constants.MATERIAL_TextureDirtyFlag);
+                            }
+                        }
+                        if (cdfGenerator) {
+                            cdfGenerator.iblSource = this;

```

Comment:
Let me see if I've got this right:
* we first set `cdfGenerator.iblSource` to the original cube texture data (not filtered)
* we call `hdrIrradianceFiltering.prefilter` to generate the irradiance map. This process will use the cdf texture generated by the step above
* we call `hdrFiltering.prefilter`, which will replace the cube data by prefiltered data
* we set `cdfGenerator.iblSource` to the current cube texture, so a texture with prefiltered data

Is this correct? I want to make sure that the generation of the cdf map from the raw data in step 1 and the filtered data in step 4 corresponds to what is expected.

Also, I'm not sure it's the duty of `HDRCubeTexture` to define `cdfGenerator.iblSource` at the end of the process. If, for example, we call `HDRCubeTexture` several times in a row to convert several textures, we'll be generating cdf maps unnecessarily. Even if called only once, the prefiltered cube may not be the one the user will use as the environment map of the scene, so setting `iblSource` is not necessary in this case either.

Reply:
Yes, those are the steps.

Good point about setting `iblSource`